tests/flow/test_intern_string.py

```diff
@@ -0,0 +1,429 @@
+
+        assertStringPoolStats(self.conn, 1, 1)
+
+        # make sure string is valid
+        value = self.graph.query("MATCH (n) RETURN n.value").result_set[0][0]
+        self.env.assertEquals(value, SMALL_STRING)
+
+        # delete the last graph in the DB
+        self.graph.delete()
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 0, 0)
+
+    def test_nested_shared_string(self):
+        # make sure shared string via containers isn't released prematurely
+
+        # create a node with a string attribute
+        p = {'s': SMALL_STRING}
+        self.graph.query("CREATE (:A {v:intern($s)}), (:B {v:[intern($s)]})", p)
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 1, 2)
+
+        # delete first reference to shared string
+        res = self.graph.query("MATCH (a:A) DELETE a")
+        self.env.assertEquals(res.nodes_deleted, 1)
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 1, 1)
+
+        # verify that string is still valid
+        res = self.graph.query("MATCH (b:B) RETURN b.v[0]").result_set
+        self.env.assertEquals(res[0][0], SMALL_STRING)
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 1, 1)
+
+        self.graph.delete()
+        assertStringPoolStats(self.conn, 0, 0)
+
+        #-----------------------------------------------------------------------
+
+        # same node containing the same string multiple times
+        self.graph.query("CREATE (:A {a: intern($s), b:[intern($s)], c:intern($s)})", p)
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 1, 3)
+
+        # delete first reference
+        self.graph.query("MATCH (a:A) SET a.a = NULL")
+        res = self.graph.query("MATCH (a:A) RETURN a.b[0], a.c").result_set
+        self.env.assertEquals(res[0][0], SMALL_STRING)
+        self.env.assertEquals(res[0][1], SMALL_STRING)
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 1, 2)
+
+        # delete second reference
+        self.graph.query("MATCH (a:A) SET a.b = NULL")
+        res = self.graph.query("MATCH (a:A) RETURN a.c").result_set
+        self.env.assertEquals(res[0][0], SMALL_STRING)
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 1, 1)
+
+        # delete last reference & delete node
+        self.graph.query("MATCH (a:A) SET a.c = NULL")
+        self.graph.query("MATCH (a:A) DELETE a")
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 0, 0)
+
+    def test_intermediate_intern_string(self):
+        # make sure intermediate intern string are removed
+
+        # create a node with a string attribute
+        self.graph.query("WITH intern('ABCDEF') AS intermediate RETURN intermediate")
+
+        # validate string pool stats
+        assertStringPoolStats(self.conn, 0, 0)
+
+    def test_undolog(self):
```

Comment:
missing test create node with string that interned that failed see that the string is removed from the string poll